modules/core/src/main/scala-3/com/siriusxm/snapshot4s/ReprForAdt.scala

```diff
@@ -31,9 +31,60 @@ See the guide for a list of supported types:
 https://siriusxm.github.io/snapshot4s/inline-snapshots/#supported-data-types""")
     }
 
-  inline def derived[A](using m: Mirror.Of[A]): Repr[A] =
-    // we don't care about the actual instances
-    // summon is called to make sure all member types
-    // also have their Repr instances
-    summonAll[Tuple.Map[m.MirroredElemTypes, Repr]]
-    Repr.fromPprint
+  inline def derived[A](using m: Mirror.Of[A]): Repr[A] = {
+    val elemInstances = summonAll[Tuple.Map[m.MirroredElemTypes, Repr]]
+
+    inline m match {
+      case s: Mirror.SumOf[A]     => derivedSum(s, elemInstances)
+      case p: Mirror.ProductOf[A] => derivedProduct(p, elemInstances)
+    }
+  }
+
+  private inline def derivedProduct[A](
+      mirror: Mirror.ProductOf[A],
+      elemInstances: Tuple
+  ): Repr[A] = {
+    val typeName   = constValue[mirror.MirroredLabel]
+    val elemLabels = constValueTuple[mirror.MirroredElemLabels]
+    ProductRepr(typeName, elemLabels, elemInstances)
+  }
+
+  private inline def derivedSum[A](mirror: Mirror.SumOf[A], elemInstances: Tuple): Repr[A] =
+    SumRepr(mirror, elemInstances)
+
+  class ProductRepr[A](typeName: String, elemLabels: Tuple, elemInstances: Tuple) extends Repr[A] {
+
+    def toSourceString(a: A): String = {
+      val product  = a.asInstanceOf[Product]
+      val elements = product.productIterator.toArray
+
+      if elements.isEmpty then typeName
+      else {
+        // Convert tuple elements to array safely
+        val reprInstancesArray = elemInstances.toArray
+        val elementReprs       = elements.zipWithIndex.map { case (elem, idx) =>
```

Comment:
Instead of zipping with the index, how about we `zip` the collections together?

```
        val reprInstances = elemInstances.toList.map(_.asInstanceOf[Repr[Any]])
        val labels        = elemLabels.toList.map(_.asInstanceOf[String])
        val namedArgs     =
          elements.zip(reprInstances).zip(labels).map { case ((elem, reprInstance), label) =>
            val repr = reprInstance.toSourceString(elem)
            s"$label = $repr"
          }
        s"$typeName(${namedArgs.mkString(", ")})"
```